time_integration/one_step_theta.h

```diff
@@ -0,0 +1,206 @@
+        pcout << "average iterations " << (double)total_iterations / ((double)counter_) << std::endl;
+    }
+
+private:
+    void create_right_hand_side(bool const zero_out) const;
+
+    void local_apply_inverse_mass_matrix(const MatrixFree<dim, Number> &data,
+                                         LinearAlgebra::distributed::Vector<Number> &dst,
+                                         const LinearAlgebra::distributed::Vector<Number> &src,
+                                         const std::pair<unsigned int, unsigned int> &cell_range) const;
+
+    Operator &pde_operator_;
+
+    double const Theta_;
+
+    mutable double dt_;
+    mutable double old_time_;
+
+    mutable LinearAlgebra::distributed::Vector<Number> old_solution_;
+    mutable LinearAlgebra::distributed::Vector<Number> right_hand_side_;
+    const LevelSetProblemParameters &param_;
+
+    ConditionalOStream pcout;
+    mutable int min_iterations = std::numeric_limits<double>::infinity();
+    mutable int total_iterations = 0;
+    mutable int max_iterations = 0;
+    mutable int counter_ = 0;
+
+
+};
+
+template <typename Operator, int dim, int fe_degree>
+OneStepTheta<Operator, dim, fe_degree>::OneStepTheta(Operator &pde_operator,
+                                                     const LevelSetProblemParameters &param,
+                                                     LinearAlgebra::distributed::Vector<Number> const &) : pde_operator_(pde_operator),
+                                                                                                           param_(param),
+                                                                                                           pcout(std::cout, Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0),
+                                                                                                           Theta_(param.Theta_advection) {}
+
+template <typename Operator, int dim, int fe_degree>
+OneStepTheta<Operator, dim, fe_degree>::OneStepTheta(Operator &pde_operator,
+                                                     const LevelSetProblemParameters &param) : pde_operator_(pde_operator),
+                                                                                                 param_(param),
+                                                                                                 pcout(std::cout, Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) == 0),
+                                                                                                 Theta_(param.Theta_IMEX) {}
+
+/* Left hand operator of the solver*/
+template <typename Operator, int dim, int fe_degree>
+void OneStepTheta<Operator, dim, fe_degree>::vmult(LinearAlgebra::distributed::Vector<Number> &dst,
+                                                   const LinearAlgebra::distributed::Vector<Number> &src) const
+{
+
+
+    dst = 0;
+    pde_operator_.apply_operator(old_time_ + dt_, dst, src);
```

Comment:
For fixing the dirichlet BC issue, I would introduce a new function
```suggestion
    pde_operator_.apply_operator_with_homogeneous_DBC(old_time_ + dt_, dst, src);
```